packages/backend-app-api/src/services/implementations/httpAuth/httpAuthServiceFactory.ts

```diff
@@ -41,128 +39,205 @@ function getTokenFromRequest(req: Request) {
-      await this.#getCredentials(req),
-    );
+    // Limited and full credentials are treated as two separate cases, this lets
+    // us avoid internal dependencies between the AuthService and
+    // HttpAuthService implementations
+    const credentials = options?.allowLimitedAccess
+      ? await this.#getLimitedCredentials(req)
+      : await this.#getCredentials(req);
+
+    const allowed = options?.allow;
+    if (!allowed) {
+      return credentials as any;
+    }
 
-    const allowedPrincipalTypes = options?.allow;
-    const allowedAuthMethods: Array<'token' | 'cookie' | 'none'> =
-      options?.allowedAuthMethods ?? ['token'];
+    if (this.#auth.isPrincipal(credentials, 'none')) {
+      if (allowed.includes('none' as TAllowed)) {
+        return credentials as any;
+      }
+
+      throw new AuthenticationError('Missing credentials');
+    } else if (this.#auth.isPrincipal(credentials, 'user')) {
+      if (allowed.includes('user' as TAllowed)) {
+        return credentials as any;
+      }
 
-    if (
-      credentials.authMethod !== 'none' &&
-      !allowedAuthMethods.includes(credentials.authMethod)
-    ) {
       throw new NotAllowedError(
-        `This endpoint does not allow the '${credentials.authMethod}' auth method`,
+        `This endpoint does not allow 'user' credentials`,
       );
-    }
-
-    if (
-      allowedPrincipalTypes &&
-      !allowedPrincipalTypes.includes(credentials.principal.type as TAllowed)
-    ) {
-      if (credentials.authMethod === 'none') {
-        throw new AuthenticationError();
+    } else if (this.#auth.isPrincipal(credentials, 'service')) {
+      if (allowed.includes('service' as TAllowed)) {
+        return credentials as any;
       }
+
       throw new NotAllowedError(
-        `This endpoint does not allow '${credentials.principal.type}' credentials`,
+        `This endpoint does not allow 'service' credentials`,
       );
     }
 
-    return credentials as any;
+    throw new NotAllowedError(
+      'Unknown principal type, this should never happen',
+    );
   }
 
-  async issueUserCookie(res: Response): Promise<void> {
-    const credentials = await this.credentials(res.req, { allow: ['user'] });
+  async issueUserCookie(
+    res: Response,
+    options?: { credentials?: BackstageCredentials },
+  ): Promise<{ expiresAt: Date }> {
+    let credentials: BackstageCredentials<BackstageUserPrincipal>;
+    if (options?.credentials) {
+      if (!this.#auth.isPrincipal(options.credentials, 'user')) {
+        throw new AuthenticationError(
+          'Refused to issue cookie for non-user principal',
+        );
+      }
+      credentials = options.credentials;
+    } else {
+      credentials = await this.credentials(res.req, { allow: ['user'] });
+    }
+
+    const existingExpiresAt = await this.#existingCookieExpiration(res.req);
+    if (
+      existingExpiresAt &&
+      existingExpiresAt.getTime() < Date.now() - FIVE_MINUTES_MS
```

Comment:
```suggestion
      existingExpiresAt.getTime() - FIVE_MINUTES_MS > Date.now()
```

right?

alternatively

```suggestion
      existingExpiresAt.getTime() > Date.now() + FIVE_MINUTES_MS
```

Reply:
yep extracted for bonus clarity 😅